File: coinffeine-common-test/src/main/java/com/coinffeine/common/NetworkTestUtils.java
package com.coinffeine.common;

import java.io.IOException;
import java.net.ServerSocket;
import java.util.LinkedList;
import java.util.List;

/**
 * Utilities for integration tests using network facilities.
 */
public class NetworkTestUtils {

    /**
     * Find a number of free TCP ports.
     *
     * Ports are found by opening probe server sockets.
     *
     * @param n Number of ports to find
     * @return  List of available ports
     * @throws RuntimeException When probe server sockets cannot be created or closed.
     */
    public static List<Integer> findAvailableTcpPorts(int n) {
        List<ServerSocket> successfulSockets = new LinkedList<>();
        while (successfulSockets.size() < n) {
            try {
                ServerSocket socket = new ServerSocket(0);
                successfulSockets.add(socket);
            } catch (Exception ex) {
                throw new RuntimeException("Cannot open probe socket", ex);
            }
        }
        List<Integer> availablePorts = new LinkedList<>();
        for (ServerSocket socket : successfulSockets) {
            availablePorts.add(socket.getLocalPort());
            try {
                socket.close();
            } catch (IOException ex) {
                throw new RuntimeException("Cannot close probe server socket", ex);
            }
        }
        return availablePorts;
    }
}

File: coinffeine-common-test/src/main/scala/com/coinffeine/common/NetworkTestUtils.scala
package com.coinffeine.common

import java.net.ServerSocket
import java.io.IOException
import scala.util.Try

object NetworkTestUtils {

  /** Find a number of free TCP ports.
    *
    * Ports are found by opening probe server sockets.
    *
    * @param n Number of ports to find
    * @return  List of available ports
    * @throws RuntimeException When probe server sockets cannot be created or closed.
    */
  def findAvailableTcpPorts(n: Int): Seq[Int] = {
    val successfulSockets = Stream.continually(try {
      new ServerSocket(0)
    } catch {
      case ex: Exception => throw new RuntimeException("Cannot open probe socket", ex)
    }).take(n).toSeq
    successfulSockets.foreach(_.close)
    successfulSockets.map(_.getLocalPort)
  }
}
